src/environmentd/src/http/sql.rs

```diff
@@ -453,25 +454,49 @@ pub struct CommandStarting {
 /// accumulate into a Vec and send all at once. WebSocket clients send each
 /// message as they occur.
 #[async_trait]
-trait ResultSender {
+trait ResultSender: Send {
     /// Adds a result to the client. Returns Err if sending to the client
     /// produced an error and the server should disconnect. Returns Ok(Err) if
     /// the statement produced an error and should error the transaction, but
     /// remain connected. Returns Ok(Ok(())) if the statement succeeded.
-    async fn add_result(&mut self, res: StatementResult) -> Result<Result<(), ()>, anyhow::Error>;
-    /// Awaits a future while also able to check the status of the client
-    /// connection. Should return an error if the client connection has gone
-    /// away.
-    async fn await_rows<F, R>(&mut self, f: F) -> Result<R, anyhow::Error>
+    async fn add_result<C, F>(
+        &mut self,
+        canceled: C,
+        res: StatementResult,
+    ) -> Result<Result<(), ()>, anyhow::Error>
     where
-        F: Future<Output = R> + Send;
+        C: Fn() -> F + Send + Sync,
+        F: Future<Output = ()> + Send;
+    /// A future that checks the status of the client connection. Returns an error if the client
+    /// connection has gone away.
+    fn connection_error(&mut self) -> BoxFuture<anyhow::Error>;
```

Comment:
can you leave a comment here indicating that the returned future never resolves as long as the connection is alive? At first I was surprised to see that it returns **only** an `anyhow::Error` instead of something like a `Result<(), anyhow::Error>`, but then I realized that the returned Future must only resolve if the connection closes